### ISSUE TITLE
Dynamic dim issues for scatter_ and one_hot

### ISSUE THREAD

**Issue by Alex-Vasile** [iree-org/iree#21028]
### What happened?

Exporting a `scatter_` or `one_hot` op when only **some** of the input dims are marked as dynamic causes numeric issues and produces incorrect results. I've included two different examples of the same issue occurring, once with a `scatter_` op and again with a `one_hot` op. I am not sure exactly which part of each invocation goes wrong, but they both share the same common failure.

The `scatter_` mlirs performs:
```Python
router_scores1 = torch.zeros([num_tokens, num_experts], dtype=h.dtype)
router_scores1 = router_scores1.scatter_(
    1, top_experts_index, expert_gate
).transpose(0, 1).T
```

The `one_hot` mlirs performs:
```Python
one_hot_expert_indices = F.one_hot(
    top_experts_index, num_classes=num_experts
).to(dtype=h.dtype, device=h.device)
weighted_scores = one_hot_expert_indices * expert_gate.unsqueeze(-1)
router_scores = weighted_scores.sum(dim=1).transpose(0, 1).T
```
For both cases, the dims of the args are:
```
dynamic_shapes = {
    "ffn_input": {0: num_tokens, 1: input_feature_dim},
    "top_experts_index": {0: num_tokens, 1: num_top_experts},
    "expert_gate": {0: num_tokens, 1: num_top_experts},
}
```

When compared against an eager execution of the code, the result from the compiled versions is incorrect if dynamic dims are used for all except the `num_top_experts` dim (I have not tested if other combinations of some static and some dynamics causes issues, though they might). Having either all or none of the dims as dynamic produces the correct result. The `num_top_experts` dim is static when compiled as part of a large model.

[one_hot_good.mlir.txt](https://github.com/user-attachments/files/20620580/one_hot_good.mlir.txt)
[one_hot_bad.mlir.txt](https://github.com/user-attachments/files/20620581/one_hot.mlir.txt)
[scatter_good.mlir.txt](https://github.com/user-attachments/files/20620579/scatter_good.mlir.txt)
[scatter_bad.mlir.txt](https://github.com/user-attachments/files/20620578/scatter_bad.mlir.txt)

I am also attaching a [zip](https://github.com/user-attachments/files/20620668/repro.zip) with a python repro script and saved inputs and correct output. The script includes the export, compilation, and run code.

### Steps to reproduce your issue

Compiling with:
```
iree-compile case.mlir \
  -o=result.vmfb \
  --iree-hal-target-device=hip \
  --iree-hip-target=gfx942 \
  --iree-opt-level=O3 \
  --iree-hal-indirect-command-buffers=true \
  --iree-stream-resource-memory-model=discrete \
  --iree-hal-memoization=true
```

### Version information

`iree-3.5.0rc20250605`

**Comment by IanWood1** [comment]
I ran the `scatter_*` mlir file. Looking at the output of [--iree-flow-trace-dispatch-tensors](https://gist.github.com/IanWood1/f09d46d2b00f9bbece88ed5348074865) the first dispatch, `async_dispatch_0_elementwise_broadcast_D_i32xi32xf32`, seems to be the culprit. That is where the outputs first differ. Here are the good/bad dispatches: https://gist.github.com/IanWood1/b61bb257573e5bcf3064df865ce24df0.

The scatter dispatches were nearly identical so I don't think they are the issue.

Update: I ran the same comparison with the llvmcpu backend and both `scatter_good.mlir` and `scatter_bad.mlir` produced the correct results. I think this points to a GPU codegen issue (VectorDistribution).

**Comment by IanWood1** [comment]
`FoldCollapseShapeIntoInterfaceTensorStoreFullSlice` has a bug where it will fold

```mlir
%collapsed_8 = tensor.collapse_shape %48 [[0, 1]] : tensor<?x2xi32> into tensor<?xi32>
iree_tensor_ext.dispatch.tensor.store %collapsed_8, %33, offsets = [0, 0], sizes = [%30, 1], strides = [1, 1] : tensor<?xi32> -> !iree_tensor_ext.dispatch.tensor<readwrite:tensor<?x2xi32>>{%30}
```

into

```mlir
iree_tensor_ext.dispatch.tensor.store %50, %34, offsets = [0, 0], sizes = [%33, 2], strides = [1, 1] : tensor<?x2xi32> -> !iree_tensor_ext.dispatch.tensor<readwrite:tensor<?x2xi32>>{%33}
```
